gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java

```diff
@@ -3649,8 +3652,28 @@ protected final void releaseAfterGetInitialImageLatch() {
           + getFullPath());
     }
     releaseLatch(this.initializationLatchAfterGetInitialImage);
+
+    writeUnlockEnqueueDelta();
+  }
+
+  protected final void readLockEnqueueDelta() {
+    deltaLock.readLock().lock();
+  }
+
+  protected final void readUnlockEnqueueDelta() {
+    deltaLock.readLock().unlock();
   }
 
+  protected final void writeLockEnqueueDelta() {
+    if (!deltaLock.isWriteLocked())
+      deltaLock.writeLock().lock();
+  }
+
+  protected final void writeUnlockEnqueueDelta() {
+    if (deltaLock.isWriteLocked()) {
+      deltaLock.writeLock().unlock();
+    }
+  }
```

Comment:
The isWriteLocked() checks have race conditions. It also will cause other threads to skip locking if already locked by another thread. Is that the required behaviour?? To me that defeats the purpose of these locks which were meant to guard multiple threads trying to enqueue concurrently (or is that the purpose?).

Reply:
>> The isWriteLocked() checks have race conditions.

Since the write lock is attempted only by the gii processor thread so for write locks there isn't any race.

Reply:
So why is this check there at all? Is it being expected that write lock may not have been taken at this point? As I mentioned that such a case would be a bug and so this check should be removed.

Reply:
I think since the releaseAfterGetInitialImageLatch() will be called both in case of success/failure(exception somewhere in between gii process) so this check would have been required, otherwise blindly releasing it was fine.
BTW @suranjan we can perhaps have a read write lock just for delta scenario and write lock/unlock around requestUnappliedDeltas method and instead of region init flag, check delta fetched flag or something in the basic putter thread.